### PR TITLE
Non-trivial merge into 0.23

### PR DESCRIPTION
**Pull request by rossabaker**

### CODE REVIEW

**Comment by rossabaker** [comment]
This isn't working for me locally, and I'm perplexed.  I suspect it's related to the JVM/JS split that begins in 0.23.  I had to add a bunch of `rootJVM` to the lint aliases.

```
[error] (client / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] (core / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] (circe / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] (ember-client / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] (boopickle / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] (jawn / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] (tests / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] (laws / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] (ember-core / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] (testing / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] (dsl / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] (rootJVM / scalafixAll) scalafix.sbt.InvalidArgument: 2 errors
[error] [E0] Unknown rule 'Http4sGeneralLinters'
[error] [E1] Unknown rule 'Http4sUseLiteralsSyntax'
[error] Total time: 35 s, completed Nov 2, 2021, 12:39:30 AM
```

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -85,6 +85,10 @@ jobs:
       - name: Compile
         run: 'sbt ++${{ matrix.scala }} ''${{ matrix.ci }}'' test:compile'
 
+      - name: Check Scalafix rules
+        if: matrix.scala != '3.0.2'
```
Comment (armanbilge) [inline .github/workflows/ci.yml:89]: Not sure where this is sourcing from but should be 3.1.0 maybe?

**Comment by bplommer** [comment]
I'm so, so sorry 😱

**Comment by bplommer** [comment]
I've pushed (what I hope is) a fix for the scalafix issues - let's see what CI thinks.

File: build.sbt [hunk 2]
```diff
@@ -894,6 +961,7 @@ def http4sCrossProject(name: String, crossType: CrossType) =
     )
     .enablePlugins(Http4sPlugin)
     .jsConfigure(_.disablePlugins(DoctestPlugin))
+    .transform(_.dependsOn(scalafixInternalRules % ScalafixConfig))
```
Comment (armanbilge) [inline build.sbt]:
```suggestion
    .dependsOn(scalafixInternalRules % ScalafixConfig)
```

Comment (bplommer) [reply build.sbt]: You can't use `dependsOn` directly there because it's a cross project - it needs `configure`.

Comment (armanbilge) [reply build.sbt]: `dependsOn` works for cross projects, see any of the cross projects above. But, maybe not with scalafix :)

Comment (bplommer) [reply build.sbt]: Only when depending on another cross-project I think - here it's depending on a plain jvm-only `Project`(which is allowed with scalafix)

**Comment by bplommer** [comment]
The two test failures are flakes - a Blaze timeout and a bad generator output.

**Review by bplommer — APPROVED** [review]

File: blaze-client/src/main/scala/org/http4s/blaze/client/Http1Connection.scala [hunk 3]
```diff
@@ -18,27 +18,37 @@ package org.http4s
...
+import java.nio.ByteBuffer
+import java.util.concurrent.TimeoutException
+import java.util.concurrent.atomic.AtomicReference
 import scala.annotation.tailrec
-import scala.concurrent.{ExecutionContext, Future}
-import scala.util.{Failure, Success}
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.Success
```
Comment (danicheg) [inline blaze-client/src/main/scala/org/http4s/blaze/client/Http1Connection.scala:51]: Sorry for being late on this, but does flattened imports are actually beneficial?

Comment (armanbilge) [reply blaze-client/src/main/scala/org/http4s/blaze/client/Http1Connection.scala:51]: FWIW I like them 👍

Btw big thanks to @bplommer for fixing this up, the messy imports were bugging me!

Comment (bplommer) [reply blaze-client/src/main/scala/org/http4s/blaze/client/Http1Connection.scala:51]: Flattened imports aren’t pretty but they reduce merge conflicts.

Comment (danicheg) [reply blaze-client/src/main/scala/org/http4s/blaze/client/Http1Connection.scala:51]: If it's helping in merging 0.22 -> 0.23 -> main, then it's worth the hassle :+1:

Comment (rossabaker) [reply blaze-client/src/main/scala/org/http4s/blaze/client/Http1Connection.scala:51]: I didn't like the expanded imports until I considered the merge angle, and then I liked them a lot.

**Pull request merged**